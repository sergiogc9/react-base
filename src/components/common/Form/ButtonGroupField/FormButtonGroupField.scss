@import "@src/scss/_globals.scss";
@import "@src/scss/_mixins.scss";
@import "@src/scss/_variables.scss";

.form-button-group-field-wrapper {
	.form-button-group-label{
		font-size: 1.2rem;
	}

	.form-button-group-field-buttons {
		margin-top: 1.25rem;

		.form-button-group-field-button {
			font-size: 1.2rem;
			text-transform: none;
			padding: 0.8rem 1.6rem;

			&.selected {
				background-color: $launchmetrics-green;
				border: 1px solid $launchmetrics-green;
			}

			&:not(.selected){
				border: 1px solid $md-grey-300;
			}

			@include respond(tab-land) {
				height: 3.2rem;
			}
		}
	}
}
