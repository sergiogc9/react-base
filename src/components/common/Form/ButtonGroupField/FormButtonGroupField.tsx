import React from 'react';
import { Button } from 'react-md';
import find from 'lodash/find';

import { ButtonGroupFieldElement } from '@src/types/form';
import { formFieldBaseValidation } from '@src/components/common/Form/Form';
import { useForceFieldValue } from '@src/lib/hooks';

import './FormButtonGroupField.scss';

export type ComponentProps = {
	element: ButtonGroupFieldElement,
	forceValue?: any,
	onChangeOption: (elementId: string, value: string, isValid: boolean) => void
};

const __validateButtonGroupField = (element: ButtonGroupFieldElement, newValue: any) => {
	return formFieldBaseValidation(element, newValue);
};

const FormButtonGroupField = (props: ComponentProps) => {
	const { element, forceValue } = props;
	const { onChangeOption } = props;

	const defaultOption = element.defaultValue ? find(element.options, { value: element.defaultValue }) : null;

	const [value, setValue] = React.useState(defaultOption ? defaultOption.value : element.options[0].value);
	useForceFieldValue(forceValue, null, element, setValue, onChangeOption, __validateButtonGroupField);

	const clickHandler = React.useCallback((event: React.MouseEvent<HTMLElement>) => {
		const id = event.currentTarget.id.replace(element.id + '-', '');
		const option = find(element.options, { id });
		setValue(option!.value);
		onChangeOption(element.id, option!.value, __validateButtonGroupField(element, option!.value));
	}, [element, onChangeOption]);

	return (
		<div id={`formButtonGroupField-${element.id}`} className={`form-button-group-field-wrapper ${element.className || ''}`}>
			{element.label ? <span className="form-button-group-label">{element.label}</span> : null}
			<div className="form-button-group-field-buttons">
				{element.options.map(choice =>
					<Button
						id={`${element.id}-${choice.id}`}
						key={`form-button-group-${choice.id}`}
						flat
						className={`form-button-group-field-button ${value === choice.value ? 'selected' : ''}`}
						onClick={clickHandler}
						{...choice.buttonProps}
					>
						{choice.label}
					</Button>
				)}
			</div>
		</div>
	);
};

export default React.memo(FormButtonGroupField);
