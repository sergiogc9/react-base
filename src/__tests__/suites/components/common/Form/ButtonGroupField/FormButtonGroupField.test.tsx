import React from "react";
import { mount } from "enzyme";
import { act } from 'react-dom/test-utils';

import FormButtonGroupField from '@src/components/common/Form/ButtonGroupField/FormButtonGroupField';
import { ComponentProps } from '@src/components/common/Form/ButtonGroupField/FormButtonGroupField';
import TestHelper from "@src/__tests__/utils/Helper/TestHelper";

const element = TestHelper.buttonGroupElement;

let onChangeMock = jest.fn();

describe("Button group field component", () => {
	let wrapper: any;

	const mockClickEvent = (id: string) => ({
		currentTarget: {
			id
		}
	});

	const updateComponent = (props: Partial<ComponentProps> = {}) => {
		act(() => {
			wrapper = mount(
				<FormButtonGroupField
					element={element}
					onChangeOption={onChangeMock}
					{...props}
				/>);
		});
	};

	beforeEach(() => {
		onChangeMock = jest.fn();
		updateComponent();
	});

	it("Component default snapshot", () => {
		expect(wrapper.html()).toMatchSnapshot();
	});

	it("Component without label", () => {
		updateComponent({ element: { ...element, label: '' } });
		expect(wrapper.find('.form-button-group-intro')).toHaveLength(0);
	});

	it("Component without default value", () => {
		updateComponent({ element: { ...element, defaultValue: '' } });
		expect(wrapper.html()).toMatchSnapshot();
	});

	it("Component second option as default value", () => {
		updateComponent({ element: { ...element, defaultValue: element.options[1].value } });
		expect(wrapper.html()).toMatchSnapshot();
	});

	it("Component first option click", () => {
		act(() => wrapper.find(".form-button-group-field-button").at(0).props().onClick(mockClickEvent(element.options[0].id)));
		expect(onChangeMock).toHaveBeenCalledTimes(1);
		expect(wrapper.html()).toMatchSnapshot();
	});

	it("Component second option click", () => {
		act(() => wrapper.find(".form-button-group-field-button").at(1).props().onClick(mockClickEvent(element.options[1].id)));
		expect(onChangeMock).toHaveBeenCalledTimes(1);
		expect(wrapper.html()).toMatchSnapshot();
	});

	it("Component on validation", () => {
		updateComponent({ element: { ...element, validations: { fn: () => false } } });
		act(() => wrapper.find(".form-button-group-field-button").at(0).props().onClick(mockClickEvent(element.options[0].id)));
		expect(onChangeMock).toHaveBeenCalledWith(element.id, element.options[0].value, false);
	});
});
