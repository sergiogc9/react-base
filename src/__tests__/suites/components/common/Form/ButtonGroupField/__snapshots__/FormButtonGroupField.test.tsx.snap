// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button group field component Component default snapshot 1`] = `"<div id=\\"formButtonGroupField-field_id\\" class=\\"form-button-group-field-wrapper \\"><span class=\\"form-button-group-label\\">Some intro</span><div class=\\"form-button-group-field-buttons\\"><button id=\\"field_id-FirstButtonId\\" type=\\"button\\" class=\\"md-btn md-btn--flat md-btn--text md-pointer--hover md-text md-inline-block form-button-group-field-button selected\\"><div class=\\"md-ink-container\\"></div>First choice</button><button id=\\"field_id-Second ButtonId\\" type=\\"button\\" class=\\"md-btn md-btn--flat md-btn--text md-pointer--hover md-text md-inline-block form-button-group-field-button \\"><div class=\\"md-ink-container\\"></div>Second choice</button></div></div>"`;

exports[`Button group field component Component first option click 1`] = `"<div id=\\"formButtonGroupField-field_id\\" class=\\"form-button-group-field-wrapper \\"><span class=\\"form-button-group-label\\">Some intro</span><div class=\\"form-button-group-field-buttons\\"><button id=\\"field_id-FirstButtonId\\" type=\\"button\\" class=\\"md-btn md-btn--flat md-btn--text md-pointer--hover md-text md-inline-block form-button-group-field-button selected\\"><div class=\\"md-ink-container\\"></div>First choice</button><button id=\\"field_id-Second ButtonId\\" type=\\"button\\" class=\\"md-btn md-btn--flat md-btn--text md-pointer--hover md-text md-inline-block form-button-group-field-button \\"><div class=\\"md-ink-container\\"></div>Second choice</button></div></div>"`;

exports[`Button group field component Component second option as default value 1`] = `"<div id=\\"formButtonGroupField-field_id\\" class=\\"form-button-group-field-wrapper \\"><span class=\\"form-button-group-label\\">Some intro</span><div class=\\"form-button-group-field-buttons\\"><button id=\\"field_id-FirstButtonId\\" type=\\"button\\" class=\\"md-btn md-btn--flat md-btn--text md-pointer--hover md-text md-inline-block form-button-group-field-button \\"><div class=\\"md-ink-container\\"></div>First choice</button><button id=\\"field_id-Second ButtonId\\" type=\\"button\\" class=\\"md-btn md-btn--flat md-btn--text md-pointer--hover md-text md-inline-block form-button-group-field-button selected\\"><div class=\\"md-ink-container\\"></div>Second choice</button></div></div>"`;

exports[`Button group field component Component second option click 1`] = `"<div id=\\"formButtonGroupField-field_id\\" class=\\"form-button-group-field-wrapper \\"><span class=\\"form-button-group-label\\">Some intro</span><div class=\\"form-button-group-field-buttons\\"><button id=\\"field_id-FirstButtonId\\" type=\\"button\\" class=\\"md-btn md-btn--flat md-btn--text md-pointer--hover md-text md-inline-block form-button-group-field-button \\"><div class=\\"md-ink-container\\"></div>First choice</button><button id=\\"field_id-Second ButtonId\\" type=\\"button\\" class=\\"md-btn md-btn--flat md-btn--text md-pointer--hover md-text md-inline-block form-button-group-field-button selected\\"><div class=\\"md-ink-container\\"></div>Second choice</button></div></div>"`;

exports[`Button group field component Component without default value 1`] = `"<div id=\\"formButtonGroupField-field_id\\" class=\\"form-button-group-field-wrapper \\"><span class=\\"form-button-group-label\\">Some intro</span><div class=\\"form-button-group-field-buttons\\"><button id=\\"field_id-FirstButtonId\\" type=\\"button\\" class=\\"md-btn md-btn--flat md-btn--text md-pointer--hover md-text md-inline-block form-button-group-field-button selected\\"><div class=\\"md-ink-container\\"></div>First choice</button><button id=\\"field_id-Second ButtonId\\" type=\\"button\\" class=\\"md-btn md-btn--flat md-btn--text md-pointer--hover md-text md-inline-block form-button-group-field-button \\"><div class=\\"md-ink-container\\"></div>Second choice</button></div></div>"`;
